### PR TITLE
Add harm runtime

### PR DESCRIPTION
**Pull request by monoid**
Memory buffers for collecting instructions, applying relocation and execution.

Ref #31.

## Summary by CodeRabbit

* **New Features**
  * Pluggable memory backends: non-executable foreign buffers plus optional executable mmap support; new assembler pipeline to emit executable code with automatic label resolution and relocations.

* **Refactor**
  * Unified 64-bit address/offset types across relocation and memory APIs.

* **Tests**
  * Added unit and integration tests, including alignment, relocation edge cases, and executable-mmap execution checks.

* **Chores**
  * Made mmap dependency feature-optional, adjusted crate features, and added small tooling/dev dependencies.

[Review Change Stack](https://app.coderabbit.ai/change-stack/monoid/harm/pull/50)

### CODE REVIEW

**Comment by monoid** [comment]
@coderabbitai please, review.

File: harm-runtime/src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    /// Current writing position.
+    fn pos(&self) -> usize;
+
+    /// If memory has fixed capacity, return it.
+    ///
+    /// A `Vec` is not considered a memory of fixed capacity because it can grow indefinitely.
+    fn capacity(&self) -> Option<usize>;
+
+    /// Append data to the memory.  Should fail when it reaches memory's capacity.
+    fn try_extend<I: Iterator<Item = u8>>(&mut self, bytes: I) -> Result<(), Self::ExtendError>;
```
[the remark this answers is not in the document]
Comment (monoid) [reply harm-runtime/src/memory.rs]: Implementing atomicity without either additional memory or changing the bytes to be overwritten is not possible.  Though changing memory might be OK.

[the remark this answers is not in the document]
Comment (monoid) [reply harm-runtime/src/memory.rs]: Unfortunately, using slice wouldn't work: it has to allocate memory somewhere too.  The iterator abstraction is used because instructions in the `harm` package generate their bytes together with relocations, and virtual instructions like `mov` can generate different number of bytes (it may expand to `orr`, it may expand to series of `movk`/`movz`/`movn`, etc).

[the remark this answers is not in the document]
Comment (monoid) [reply harm-runtime/src/memory.rs]: ~~Using a scratch buffer in the builder would work, though.~~

[the remark this answers is not in the document]
Comment (monoid) [reply harm-runtime/src/memory.rs]: My idea is to eventually avoid `alloc` completely, so it has to be a fixed-size buffer in the `Assembler`.  But it would require large enough to contain all the virtual commands...  On the other hand, it contradicts with a planned user macro feature.

**Comment by monoid** [comment]
@coderabbitai please, review.
